libqtile/widget/openweather.py

```diff
@@ -0,0 +1,215 @@
+            None,
+            'Open Weather access key'
+        ),
+        (
+            'woeid',
+            None,
+            """ID of city. Can be looked up from:
+            https://openweathermap.org/find
+            Takes precedence over location and coordinates."""
+        ),
+        (
+            'location',
+            None,
+            """Name of the city. Country name can be appended
+            like cambridge,NZ. Takes precedence over coordinates."""
+        ),
+        (
+            'coordinates',
+            None,
+            """Dictionary containing latitude and longitude
+               Example: coordinates={"longitude": "77.22",
+                                     "latitude": "28.67"}"""
+        ),
+        (
+            'format',
+            '{location}: {temp} {temp_units} {humidity} {weather_details}',
+            'Display format'
+        ),
+        ('metric', True, "True to use metric (Default). Imperial if false"),
+        ('language', 'en',
+         """Language of response. List of languages supported can
+         be seen at: https://openweathermap.org/current under
+         Multilingual support""")
+    ]
+
+    def __init__(self, **config):
+        GenPollUrl.__init__(self, **config)
+        self.add_defaults(OpenWeather.defaults)
+
+    @property
+    def url(self):
+        if not self.app_key and not self.location and not self.coordinates:
+            return None
+
+        params = {}
+        if self.woeid:
+            params['id'] = self.woeid
+        elif self.location:
+            params['q'] = self.location
+        elif self.coordinates:
+            params['lat'] = self.coordinates['latitude']
+            params['lon'] = self.coordinates['longitude']
+
+        params['appid'] = self.app_key
+
+        if self.metric:
+            params['units'] = "metric"
+        else:
+            params['units'] = "imperial"
+
+        params['lang'] = self.language
+
+        return QUERY_URL + urlencode(params)
+
+    def parse(self, response):
+        try:
+            opwr = OpenWeatherResponseParser(response)
+        except OpenWeatherResponseError as e:
+            return "Error {}".format(e.resp_code)
+
+        data = {}
+        if self.metric:
+            data['temp_units'] = "\u00B0" + "c"
+            data['wind_speed_units'] = "kmph"
+        else:
+            data['temp_units'] = " F"
+            data['wind_speed_units'] = "mph"
+
+        data['location'] = opwr.get_location_name()
+        data['temp'] = opwr.get_actual_temperature()
+        data['humidity'] = str(opwr.get_humidity()) + "%"
+        data['weather_details'] = opwr.get_weather_details()
+        data['sunrise'] = "\u2600 {}".format(opwr.get_sunrise_time())
+        data['sunset'] = "\u26ed {}".format(opwr.get_sunset_time())
+        data['weather_concise'] = opwr.get_weather_title()
+        data['wind_speed'] = str(opwr.get_wind_speed())
+        data['wind_direction'] = opwr.get_wind_direction()
```

Comment:
Can we make all of this simpler and assign directly the values from the answer of OWM? I think we would save a lot of bytes overall, feels kinda over-engineered here. :)

e.g.:
```python
data["location"] = response["name"]
```

Reply:
There is a separate class to parse the json output. If I have to assign directly, I will have to parse these inline. That would be become messy and unmaintainable.

Reply:
I don't know which direction to lean towards. The YahooWeather-widget basically just flattens the returned json and then the user can choose the keys he wants to access. This keeps things much more flexible in case the provider changes his format or anything, but of course the access strings might be a bit technical.